File: examples/study-log-analyzer/study_log.c
/* Study Log Analyzer, C17: bounded input, checked parsing and common report. */
#include <ctype.h>
#include <errno.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_LINE 4096
#define MAX_ROWS 4096
#define MAX_TEXT 512

typedef struct {
    char date[11];
    char language[8];
    char topic[MAX_TEXT];
    unsigned minutes;
    char result[8];
} Session;
typedef struct { char name[MAX_TEXT]; unsigned long long total; } Group;

static Session sessions[MAX_ROWS];
static Session selected[MAX_ROWS];
static Group languages[3], topics[MAX_ROWS];
static size_t count, used, language_count, topic_count;
static const char *filter_language, *filter_topic, *filter_date, *filter_search;
static int date_sort;
static size_t top = 5;

static int csv_fields(const char *line, char out[5][MAX_LINE]) {
    size_t column = 0, pos = 0;
    int quoted = 0, ended = 0, start = 1;
    memset(out, 0, 5 * MAX_LINE);
    for (size_t i = 0;; i++) {
        char ch = line[i];
        if (quoted) {
            if (ch == '\0' || ch == '\n') return 0; /* multiline fields excluded */
            if (ch == '"' && line[i + 1] == '"') { ch = '"'; i++; }
            else if (ch == '"') { quoted = 0; ended = 1; continue; }
            if (pos + 1 >= MAX_LINE) return 0;
            out[column][pos++] = ch;
            continue;
        }
        if (ended && ch != ',' && ch != '\r' && ch != '\n' && ch != '\0') return 0;
        if (ch == '"') { if (!start) return 0; quoted = 1; start = 0; continue; }
        if (ch == ',') {
            if (++column >= 5) return 0;
            pos = 0; start = 1; ended = 0; continue;
        }
        if (ch == '\r' && (line[i + 1] == '\n' || line[i + 1] == '\0')) continue;
        if (ch == '\n' || ch == '\0') return column == 4;
        if (pos + 1 >= MAX_LINE) return 0;
        out[column][pos++] = ch; start = 0;
    }
}

static int date_valid(const char *text) {
    if (strlen(text) != 10 || text[4] != '-' || text[7] != '-') return 0;
    for (int i = 0; i < 10; i++) if (i != 4 && i != 7 && !isdigit((unsigned char)text[i])) return 0;
    int y = (text[0]-'0')*1000+(text[1]-'0')*100+(text[2]-'0')*10+text[3]-'0';
    int m = (text[5]-'0')*10+text[6]-'0', d = (text[8]-'0')*10+text[9]-'0';
    if (y < 1 || m < 1 || m > 12) return 0;
    const int days[] = {0,31,28,31,30,31,30,31,31,30,31,30,31};
    int max = days[m] + (m == 2 && y % 4 == 0 && (y % 100 != 0 || y % 400 == 0));
    return d >= 1 && d <= max;
}

static int minutes_valid(const char *text, unsigned *out) {
    if (!*text || *text == '0') return 0;
    for (const char *c = text; *c; c++) if (!isdigit((unsigned char)*c)) return 0;
    errno = 0; char *end = NULL;
    unsigned long value = strtoul(text, &end, 10);
    if (errno || !end || *end || value < 1 || value > 10080) return 0;
    *out = (unsigned)value;
    return 1;
}

static int has_text(const char *text) {
    for (; *text; text++) if (!isspace((unsigned char)*text)) return 1;
    return 0;
}

static int load(const char *path) {
    FILE *file = fopen(path, "r");
    if (!file) { perror("오류: 파일 읽기 실패"); return 0; }
    char line[MAX_LINE], fields[5][MAX_LINE];
    if (!fgets(line, sizeof line, file) || !csv_fields(line, fields) ||
        strcmp(fields[0], "date") || strcmp(fields[1], "language") ||
        strcmp(fields[2], "topic") || strcmp(fields[3], "minutes") || strcmp(fields[4], "result")) {
        fprintf(stderr, "오류: 헤더는 date,language,topic,minutes,result여야 합니다\n"); fclose(file); return 0;
    }
    size_t number = 1;
    while (fgets(line, sizeof line, file)) {
        number++;
        if (!strchr(line, '\n') && !feof(file)) {
            fprintf(stderr, "오류: %zu행이 너무 깁니다\n", number); fclose(file); return 0;
        }
        if (!csv_fields(line, fields) || count >= MAX_ROWS ||
            !date_valid(fields[0]) ||
            (strcmp(fields[1], "C") && strcmp(fields[1], "Python") && strcmp(fields[1], "Rust")) ||
            !has_text(fields[2]) || strlen(fields[2]) >= MAX_TEXT ||
            (strcmp(fields[4], "pass") && strcmp(fields[4], "retry")) ||
            !minutes_valid(fields[3], &sessions[count].minutes)) {
            fprintf(stderr, "오류: %zu행의 CSV 또는 필드 값이 올바르지 않습니다\n", number); fclose(file); return 0;
        }
        Session *s = &sessions[count++];
        strcpy(s->date, fields[0]); strcpy(s->language, fields[1]);
        strcpy(s->topic, fields[2]); strcpy(s->result, fields[4]);
    }
    if (ferror(file)) { perror("오류: 파일 읽기 실패"); fclose(file); return 0; }
    fclose(file); return 1;
}

static int text_contains(const char *text, const char *query) {
    for (; *text; text++) {
        const char *a = text, *b = query;
        while (*a && *b && tolower((unsigned char)*a) == tolower((unsigned char)*b)) { a++; b++; }
        if (!*b) return 1;
    }
    return !*query;
}

static void add_group(Group *groups, size_t *length, const char *key, unsigned value) {
    for (size_t i = 0; i < *length; i++) if (!strcmp(groups[i].name, key)) {
        groups[i].total += value; return;
    }
    Group *g = &groups[(*length)++];
    strcpy(g->name, key); g->total = value;
}

static int group_cmp(const void *a, const void *b) {
    return strcmp(((const Group *)a)->name, ((const Group *)b)->name);
}
static int session_cmp(const void *pa, const void *pb) {
    const Session *a = pa, *b = pb;
    int order = date_sort ? strcmp(a->date, b->date) :
                (a->minutes < b->minutes ? 1 : a->minutes > b->minutes ? -1 : 0);
    if (order) return order;
    if (date_sort) {
        if ((order = strcmp(a->language, b->language))) return order;
        if ((order = strcmp(a->topic, b->topic))) return order;
        return a->minutes < b->minutes ? -1 : a->minutes > b->minutes ? 1 : 0;
    }
    if ((order = strcmp(a->date, b->date))) return order;
    if ((order = strcmp(a->language, b->language))) return order;
    return strcmp(a->topic, b->topic);
}

static int options(int argc, char **argv) {
    if (argc < 2) return 0;
    for (int i = 2; i < argc; i += 2) {
        if (i + 1 >= argc) return 0;
        const char *name = argv[i], *value = argv[i + 1];
        if (!strcmp(name, "--language")) {
            if (strcmp(value,"C") && strcmp(value,"Python") && strcmp(value,"Rust")) return 0;
            filter_language = value;
        } else if (!strcmp(name, "--topic")) filter_topic = value;
        else if (!strcmp(name, "--date")) { if (!date_valid(value)) return 0; filter_date = value; }
        else if (!strcmp(name, "--search")) filter_search = value;
        else if (!strcmp(name, "--sort")) { if (strcmp(value,"date") && strcmp(value,"minutes")) return 0; date_sort = !strcmp(value,"date"); }
        else if (!strcmp(name, "--top")) {
            unsigned parsed;
            if (!minutes_valid(value, &parsed) || parsed > 1000) return 0;
            top = parsed;
        } else return 0;
    }
    return 1;
}

int main(int argc, char **argv) {
    if (!options(argc, argv)) {
        fprintf(stderr,"오류: 사용법 study_log FILE [--language C|Python|Rust] [--topic TEXT] [--date YYYY-MM-DD] [--search TEXT] [--sort minutes|date] [--top N]\n");
        return 1;
    }
    if (!load(argv[1])) return 1;
    unsigned long long total = 0;
    unsigned min = UINT_MAX, max = 0;
    for (size_t i = 0; i < count; i++) {
        const Session *s = &sessions[i];
        if ((filter_language && strcmp(filter_language,s->language)) ||
            (filter_topic && strcmp(filter_topic,s->topic)) ||
            (filter_date && strcmp(filter_date,s->date)) ||
            (filter_search && !text_contains(s->topic,filter_search))) continue;
        selected[used++] = *s; total += s->minutes;
        if (s->minutes < min) min = s->minutes;
        if (s->minutes > max) max = s->minutes;
        add_group(languages,&language_count,s->language,s->minutes);
        add_group(topics,&topic_count,s->topic,s->minutes);
    }
    printf("sessions: %zu\ntotal_minutes: %llu\naverage_minutes: %.2f\nmin_minutes: %u\nmax_minutes: %u\nlanguage_totals:\n",
           used,total,used ? (double)total / used : 0.0,used ? min : 0,max);
    qsort(languages, language_count, sizeof languages[0], group_cmp);
    qsort(topics, topic_count, sizeof topics[0], group_cmp);
    for (size_t i=0;i<language_count;i++) printf("%s,%llu\n",languages[i].name,languages[i].total);
    puts("topic_totals:");
    for (size_t i=0;i<topic_count;i++) printf("%s,%llu\n",topics[i].name,topics[i].total);
    puts("top_sessions:");
    qsort(selected,used,sizeof selected[0],session_cmp);
    for (size_t i=0;i<used && i<top;i++) {
        const Session *s=&selected[i];
        printf("%s,%s,%s,%u,%s\n",s->date,s->language,s->topic,s->minutes,s->result);
    }
    return 0;
}

File: examples/study-log-analyzer/study_log.rs
//! Standalone Rust 2024 implementation with a small quoted-CSV parser.
use std::collections::BTreeMap;
use std::env;
use std::fs;

#[derive(Clone)]
struct Session {
    date: String,
    language: String,
    topic: String,
    minutes: u32,
    result: String,
}

fn rows(text: &str) -> Result<Vec<Vec<String>>, String> {
    let mut output = Vec::new();
    let mut record = Vec::new();
    let mut field = String::new();
    let mut quoted = false;
    let mut after_quote = false;
    let mut at_start = true;
    let mut chars = text.chars().peekable();
    while let Some(ch) = chars.next() {
        if quoted {
            if ch == '"' {
                if chars.peek() == Some(&'"') {
                    field.push('"'); chars.next();
                } else { quoted = false; after_quote = true; }
            } else { field.push(ch); }
            continue;
        }
        if after_quote && ch != ',' && ch != '\n' && ch != '\r' {
            return Err("닫힌 따옴표 뒤에 예상치 못한 문자가 있습니다".into());
        }
        match ch {
            '"' if at_start => { quoted = true; at_start = false; }
            '"' => return Err("필드 중간의 따옴표는 이스케이프해야 합니다".into()),
            ',' => { record.push(std::mem::take(&mut field)); at_start = true; after_quote = false; }
            '\n' => {
                record.push(std::mem::take(&mut field)); output.push(std::mem::take(&mut record));
                at_start = true; after_quote = false;
            }
            '\r' if chars.peek() == Some(&'\n') => {},
            '\r' => return Err("CRLF 또는 LF 줄바꿈을 사용하세요".into()),
            _ => { field.push(ch); at_start = false; }
        }
    }
    if quoted { return Err("닫히지 않은 CSV 따옴표".into()); }
    if !record.is_empty() || !field.is_empty() || !at_start {
        record.push(field); output.push(record);
    }
    Ok(output)
}

fn date_valid(s: &str) -> bool {
    let b = s.as_bytes();
    if b.len() != 10 || b[4] != b'-' || b[7] != b'-' ||
        !b.iter().enumerate().all(|(i, c)| i == 4 || i == 7 || c.is_ascii_digit()) { return false; }
    let (Ok(y), Ok(m), Ok(d)) = (s[0..4].parse::<u32>(), s[5..7].parse::<u32>(), s[8..10].parse::<u32>()) else { return false; };
    let leap = y % 4 == 0 && (y % 100 != 0 || y % 400 == 0);
    let max = match m { 1|3|5|7|8|10|12 => 31, 4|6|9|11 => 30, 2 if leap => 29, 2 => 28, _ => 0 };
    y > 0 && d >= 1 && d <= max
}

fn load(path: &str) -> Result<Vec<Session>, String> {
    let source = fs::read_to_string(path).map_err(|e| format!("파일 읽기 실패: {e}"))?;
    let mut parsed = rows(&source)?.into_iter();
    if parsed.next().as_deref() != Some(&["date".into(), "language".into(), "topic".into(), "minutes".into(), "result".into()]) {
        return Err("헤더는 date,language,topic,minutes,result여야 합니다".into());
    }
    parsed.enumerate().map(|(line, fields)| {
        if fields.len() != 5 { return Err(format!("{}행: 열이 정확히 5개여야 합니다", line+2)); }
        if !date_valid(&fields[0]) || !matches!(fields[1].as_str(), "C"|"Python"|"Rust") ||
            fields[2].trim().is_empty() || !matches!(fields[4].as_str(), "pass"|"retry") {
            return Err(format!("{}행: date/language/topic/result 오류", line+2));
        }
        let text = &fields[3];
        if text.starts_with('0') || !text.bytes().all(|c| c.is_ascii_digit()) {
            return Err(format!("{}행: minutes는 양의 정수여야 합니다", line+2));
        }
        let minutes = text.parse::<u32>().map_err(|_| format!("{}행: minutes 오류", line+2))?;
        if !(1..=10080).contains(&minutes) { return Err(format!("{}행: minutes는 10080 이하여야 합니다", line+2)); }
        Ok(Session { date: fields[0].clone(), language: fields[1].clone(), topic: fields[2].clone(), minutes, result: fields[4].clone() })
    }).collect()
}

struct Options {
    path: String, language: Option<String>, topic: Option<String>, date: Option<String>,
    search: Option<String>, sort: String, top: usize,
}

fn options() -> Result<Options, String> {
    let mut args = env::args().skip(1);
    let path = args.next().ok_or("사용법: study_log FILE [--language C|Python|Rust] [--topic TEXT] [--date YYYY-MM-DD] [--search TEXT] [--sort minutes|date] [--top N]")?;
    let mut result = Options { path, language: None, topic: None, date: None, search: None, sort: "minutes".into(), top: 5 };
    while let Some(flag) = args.next() {
        let value = args.next().ok_or(format!("{flag}에 값이 없습니다"))?;
        match flag.as_str() {
            "--language" if matches!(value.as_str(), "C"|"Python"|"Rust") => result.language = Some(value),
            "--topic" => result.topic = Some(value),
            "--date" if date_valid(&value) => result.date = Some(value),
            "--search" => result.search = Some(value),
            "--sort" if value == "date" || value == "minutes" => result.sort = value,
            "--top" => { result.top = value.parse().map_err(|_| "--top 숫자 오류")?; if !(1..=1000).contains(&result.top) { return Err("--top은 1~1000 사이여야 합니다".into()); } }
            _ => return Err(format!("알 수 없거나 잘못된 옵션: {flag} {value}")),
        }
    }
    Ok(result)
}

fn report(sessions: Vec<Session>, opts: &Options) -> String {
    let mut selected: Vec<Session> = sessions.into_iter().filter(|r| {
        opts.language.as_ref().is_none_or(|s| s == &r.language) &&
        opts.topic.as_ref().is_none_or(|s| s == &r.topic) &&
        opts.date.as_ref().is_none_or(|s| s == &r.date) &&
        opts.search.as_ref().is_none_or(|s| r.topic.to_lowercase().contains(&s.to_lowercase()))
    }).collect();
    let total: u64 = selected.iter().map(|r| u64::from(r.minutes)).sum();
    let min = selected.iter().map(|r| r.minutes).min().unwrap_or(0);
    let max = selected.iter().map(|r| r.minutes).max().unwrap_or(0);
    let mut languages: BTreeMap<&str, u64> = BTreeMap::new();
    let mut topics: BTreeMap<&str, u64> = BTreeMap::new();
    for r in &selected {
        *languages.entry(&r.language).or_default() += u64::from(r.minutes);
        *topics.entry(&r.topic).or_default() += u64::from(r.minutes);
    }
    let mut out = format!("sessions: {}\ntotal_minutes: {total}\naverage_minutes: {:.2}\nmin_minutes: {min}\nmax_minutes: {max}\nlanguage_totals:\n",
                          selected.len(), if selected.is_empty() { 0.0 } else { total as f64 / selected.len() as f64 });
    for (key, value) in languages { out.push_str(&format!("{key},{value}\n")); }
    out.push_str("topic_totals:\n");
    for (key, value) in topics { out.push_str(&format!("{key},{value}\n")); }
    out.push_str("top_sessions:\n");
    if opts.sort == "date" {
        selected.sort_by(|a,b| (&a.date,&a.language,&a.topic,a.minutes).cmp(&(&b.date,&b.language,&b.topic,b.minutes)));
    } else {
        selected.sort_by(|a,b| b.minutes.cmp(&a.minutes).then_with(|| (&a.date,&a.language,&a.topic).cmp(&(&b.date,&b.language,&b.topic))));
    }
    for r in selected.iter().take(opts.top) {
        out.push_str(&format!("{},{},{},{},{}\n",r.date,r.language,r.topic,r.minutes,r.result));
    }
    out
}

fn main() {
    let result = (|| -> Result<String, String> {
        let opts = options()?;
        Ok(report(load(&opts.path)?, &opts))
    })();
    match result { Ok(text) => print!("{text}"), Err(error) => { eprintln!("오류: {error}"); std::process::exit(1); } }
}
